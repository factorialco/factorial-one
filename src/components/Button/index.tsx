import { cva, type VariantProps } from "cva";
import React, { forwardRef, useState } from "react";
import { Pressable, Text, View } from "react-native";
import { cn } from "../../lib/utils";
import { Icon, type IconType } from "../Icon";

export const variants = [
  "default",
  "outline",
  "critical",
  "neutral",
  "ghost",
  "promote",
] as const;
export type ButtonVariant = (typeof variants)[number];

export const sizes = ["sm", "md", "lg"] as const;
export type ButtonSize = (typeof sizes)[number];

const buttonVariants = cva({
  base: "flex-row items-center justify-center rounded border-none grow-0",
  variants: {
    variant: {
      default: "bg-f1-background-accent-bold",
      outline: "bg-f1-background-inverse-secondary border border-f1-border",
      neutral: "bg-f1-background-secondary",
      critical: "bg-f1-background-secondary border border-f1-border",
      ghost: "bg-transparent",
      promote: "bg-f1-background-promote border border-f1-border-promote",
    },
    size: {
      sm: "h-6 rounded-sm",
      md: "h-8 rounded",
      lg: "h-10 rounded-md",
    },
    disabled: {
      true: "opacity-50",
      false: "",
    },
    round: {
      true: "aspect-square p-0",
      false: "gap-1 px-2 sm:px-3 lg:px-4",
    },
  },
  defaultVariants: {
    variant: "default",
    size: "md",
    disabled: false,
    round: false,
  },
});

const pressedVariants = cva({
  base: "",
  variants: {
    variant: {
      default: "bg-f1-background-accent-bold-hover",
      outline: "bg-f1-background-tertiary border-opacity-70",
      neutral: "bg-f1-background-secondary-hover",
      critical: "bg-f1-background-critical-bold border-transparent",
      ghost: "bg-f1-background-secondary-hover",
      promote: "bg-f1-background-promote-hover",
    },
  },
  defaultVariants: {
    variant: "default",
  },
});

const getIconColor = (variant: ButtonVariant, isPressed: boolean) => {
  switch (variant) {
    case "default":
      return "text-f1-icon-inverse";
    case "critical":
      return isPressed ? "text-f1-icon-inverse" : "text-f1-icon-critical-bold";
    default:
      return "text-f1-icon";
  }
};

const getIconOnlyColor = (variant: ButtonVariant, isPressed: boolean) => {
  switch (variant) {
    case "critical":
      return isPressed ? "text-f1-icon-inverse" : "text-f1-icon-critical-bold";
    case "default":
      return "text-f1-icon-inverse";
    case "outline":
    case "neutral":
    case "ghost":
    case "promote":
    default:
      return "text-f1-icon-bold";
  }
};

const getTextColorClass = (variant: ButtonVariant, isPressed: boolean) => {
  if (isPressed && variant === "critical") {
    return "text-f1-foreground-inverse";
  }

  switch (variant) {
    case "default":
      return "text-f1-foreground-inverse";
    case "critical":
      return "text-f1-foreground-critical";
    default:
      return "text-f1-foreground";
  }
};

export interface ButtonProps extends VariantProps<typeof buttonVariants> {
  label: string;
  onPress?: () => void | Promise<unknown>;
  disabled?: boolean;
  loading?: boolean;
  icon?: IconType;
  emoji?: string;
  hideLabel?: boolean;
  className?: string;
  accessibilityHint?: string;
  showBadge?: boolean;
  fullWidth?: boolean;
}

export const Button = forwardRef<View, ButtonProps>(function Button(
  {
    label,
    onPress,
    disabled = false,
    loading = false,
    icon,
    emoji,
    hideLabel = false,
    variant = "default",
    size = "md",
    round = false,
    className,
    accessibilityHint,
    showBadge = false,
    fullWidth = false,
  },
  ref,
) {
  const [isLoading, setIsLoading] = useState(false);
  const [isPressed, setIsPressed] = useState(false);

  const handlePress = async () => {
    if (!onPress || disabled || loading || isLoading) return;

    const result = onPress();

    if (result instanceof Promise) {
      setIsLoading(true);
      try {
        await result;
      } finally {
        setIsLoading(false);
      }
    }
  };

  const isDisabled = disabled || loading || isLoading;
  const accessibilityLabel = `${label}${isDisabled ? ", disabled" : ""}${loading || isLoading ? ", loading" : ""}`;
  const shouldShowPressed = isPressed && !isDisabled;

  return (
    <View className={`flex ${fullWidth ? "flex-1" : "item-start"}`}>
      <Pressable
        ref={ref}
        disabled={isDisabled}
        onPressIn={() => setIsPressed(true)}
        onPressOut={() => setIsPressed(false)}
        onPress={handlePress}
        className={cn(
          buttonVariants({
            variant,
            size,
            disabled: isDisabled,
            round: hideLabel && round,
          }),
          shouldShowPressed && pressedVariants({ variant }),
          className,
        )}
        accessibilityLabel={accessibilityLabel}
        accessibilityRole="button"
        accessibilityState={{
          disabled: isDisabled,
          busy: loading || isLoading,
        }}
        accessibilityHint={accessibilityHint}
      >
        {icon && (
          <Icon
            icon={icon}
            size={size === "sm" ? "sm" : "md"}
            className={cn(
              hideLabel && round ? undefined : "-ml-0.5",
              hideLabel && round
                ? getIconOnlyColor(variant, shouldShowPressed)
                : getIconColor(variant, shouldShowPressed),
            )}
          />
        )}
        {emoji && (
          <Text
            className={cn(
              "text-base font-medium",
              getTextColorClass(variant, shouldShowPressed),
            )}
          >
            {emoji}
          </Text>
        )}
        {!hideLabel && (
          <Text
            className={cn(
              "text-base font-medium",
              getTextColorClass(variant, shouldShowPressed),
            )}
          >
            {label}
          </Text>
        )}
      </Pressable>
      {showBadge && variant === "outline" && (
        <View
          accessibilityLabel="Notification Badge"
          className="absolute right-1.5 top-1.5 h-1.5 w-1.5 rounded-full bg-f1-icon-accent"
        />
      )}
    </View>
  );
});
