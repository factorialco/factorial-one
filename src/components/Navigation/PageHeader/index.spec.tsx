import {
  render,
  fireEvent,
  screen,
  within,
} from "@testing-library/react-native";
import React from "react";
import { PageHeader } from "./"; // Assuming index.tsx is in the same folder
import { ButtonProps } from "../../Button"; // Adjust path as necessary
// import { AppIcons } from "../../../index"; // Adjust path as necessary - Not directly used in mock after change

// Define a simple type for the mocked icon
type MockIconType = { name: string };

// Mock the Button component
jest.mock("../../Button", () => ({
  Button: jest.fn(
    ({
      label,
      onPress,
      icon,
      showBadge,
      ...rest
    }: ButtonProps & { icon?: MockIconType }) => {
      const MockButton = require("react-native").TouchableOpacity;
      const MockText = require("react-native").Text;
      return (
        <MockButton onPress={onPress} accessibilityLabel={label} {...rest}>
          <MockText>{label}</MockText>
          {icon && <MockText>Icon:{icon.name}</MockText>}
          {showBadge && <MockText>Badge</MockText>}
        </MockButton>
      );
    },
  ),
}));

// Mock AppIcons - specifically the Bell icon used by PageHeader's NotificationAction
jest.mock("../../../index", () => ({
  AppIcons: {
    Bell: { name: "Bell" } as MockIconType, // Ensure the mock matches MockIconType
  },
}));

describe("PageHeader", () => {
  const mockOnPressNotification = jest.fn();

  const defaultProps = {
    title: "Test Page Title",
  };

  const propsWithActions = {
    ...defaultProps,
    actions: [
      {
        type: "notifications" as const,
        label: "Notifications",
        onPress: mockOnPressNotification,
        showBadge: false,
      },
    ],
  };

  const propsWithMultipleActions = {
    ...defaultProps,
    actions: [
      {
        type: "notifications" as const,
        label: "Notifications",
        onPress: mockOnPressNotification,
        showBadge: true,
      },
      {
        type: "notifications" as const, // Assuming another action of the same type for simplicity
        label: "Alerts",
        onPress: jest.fn(),
      },
    ],
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  // Snapshot tests
  it("Snapshot - renders correctly with only a title", () => {
    const { toJSON } = render(<PageHeader {...defaultProps} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("Snapshot - renders correctly with one action", () => {
    const { toJSON } = render(<PageHeader {...propsWithActions} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("Snapshot - renders correctly with multiple actions", () => {
    const { toJSON } = render(<PageHeader {...propsWithMultipleActions} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("Snapshot - renders correctly with one action with a badge", () => {
    const propsWithBadge = {
      ...defaultProps,
      actions: [
        {
          type: "notifications" as const,
          label: "Notifications",
          onPress: mockOnPressNotification,
          showBadge: true,
        },
      ],
    };
    const { toJSON } = render(<PageHeader {...propsWithBadge} />);
    expect(toJSON()).toMatchSnapshot();
  });

  // Functional tests
  it("renders the title correctly", () => {
    render(<PageHeader {...defaultProps} />);
    const title = screen.getByText("Test Page Title");
    expect(title).toBeDefined();
  });

  it("does not render actions if actions prop is not provided", () => {
    render(<PageHeader {...defaultProps} />);
    const notificationButton = screen.queryByText("Notifications");
    expect(notificationButton).toBeNull();
  });

  it("renders action buttons when actions prop is provided", () => {
    render(<PageHeader {...propsWithActions} />);
    const notificationButton = screen.getByText("Notifications");
    expect(notificationButton).toBeDefined();
  });

  it("calls the onPress handler of an action button when pressed", () => {
    render(<PageHeader {...propsWithActions} />);
    const notificationButton = screen.getByText("Notifications");
    fireEvent.press(notificationButton);
    expect(mockOnPressNotification).toHaveBeenCalledTimes(1);
  });

  it("displays a badge on an action button if showBadge is true", () => {
    render(
      <PageHeader
        title="Test Title"
        actions={[
          {
            type: "notifications",
            label: "Notifications",
            onPress: jest.fn(),
            showBadge: true,
          },
        ]}
      />,
    );
    const badge = screen.getByText("Badge");
    expect(badge).toBeDefined();
  });

  it("does not display a badge if showBadge is false or undefined", () => {
    render(
      <PageHeader
        title="Test Title"
        actions={[
          {
            type: "notifications",
            label: "Notifications",
            onPress: jest.fn(),
            // showBadge is undefined here
          },
          {
            type: "notifications",
            label: "Alerts",
            onPress: jest.fn(),
            showBadge: false,
          },
        ]}
      />,
    );
    const badge = screen.queryByText("Badge");
    expect(badge).toBeNull();
  });

  it("renders multiple action buttons correctly", () => {
    render(<PageHeader {...propsWithMultipleActions} />);
    const notificationsButton = screen.getByText("Notifications");
    const alertsButton = screen.getByText("Alerts");

    expect(notificationsButton).toBeDefined();
    expect(alertsButton).toBeDefined();

    const notificationButtonContainer = screen.getByLabelText("Notifications");

    const badge = within(notificationButtonContainer).getByText("Badge");
    expect(badge).toBeDefined();
  });
});
