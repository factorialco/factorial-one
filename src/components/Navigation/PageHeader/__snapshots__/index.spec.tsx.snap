// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHeader Snapshot - renders correctly with multiple actions 1`] = `
<View
  className="flex-row items-center justify-between px-5 py-3"
>
  <Text
    className="text-3xl font-semibold text-f1-foreground"
  >
    Test Page Title
  </Text>
  <View
    className="flex-row gap-2"
  >
    <View
      accessibilityLabel="Notifications"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text>
        Notifications
      </Text>
      <Text>
        Icon:
        Bell
      </Text>
      <Text>
        Badge
      </Text>
    </View>
    <View
      accessibilityLabel="Alerts"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text>
        Alerts
      </Text>
      <Text>
        Icon:
        Bell
      </Text>
    </View>
  </View>
</View>
`;

exports[`PageHeader Snapshot - renders correctly with one action 1`] = `
<View
  className="flex-row items-center justify-between px-5 py-3"
>
  <Text
    className="text-3xl font-semibold text-f1-foreground"
  >
    Test Page Title
  </Text>
  <View
    className="flex-row gap-2"
  >
    <View
      accessibilityLabel="Notifications"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text>
        Notifications
      </Text>
      <Text>
        Icon:
        Bell
      </Text>
    </View>
  </View>
</View>
`;

exports[`PageHeader Snapshot - renders correctly with one action with a badge 1`] = `
<View
  className="flex-row items-center justify-between px-5 py-3"
>
  <Text
    className="text-3xl font-semibold text-f1-foreground"
  >
    Test Page Title
  </Text>
  <View
    className="flex-row gap-2"
  >
    <View
      accessibilityLabel="Notifications"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text>
        Notifications
      </Text>
      <Text>
        Icon:
        Bell
      </Text>
      <Text>
        Badge
      </Text>
    </View>
  </View>
</View>
`;

exports[`PageHeader Snapshot - renders correctly with only a title 1`] = `
<View
  className="flex-row items-center justify-between px-5 py-3"
>
  <Text
    className="text-3xl font-semibold text-f1-foreground"
  >
    Test Page Title
  </Text>
  <View
    className="flex-row gap-2"
  />
</View>
`;
